### ISSUE TITLE
GTFS-rt change process - define experimental consensus, limited time window

### ISSUE THREAD

**Issue by barbeau** [google/transit#109]
In https://github.com/google/transit/pull/105 we just updated the GTFS-rt change process docs to match the currently followed process as outlined by Brian Ferris in https://groups.google.com/forum/#!msg/gtfs-realtime/bWQAgGdj3f0/mdcKwR3J8xAJ.

@gcamp suggested two changes to the process, which we agreed to table until after the above PR was adopted - so now we can discuss in this issue.

First comment thread is at https://github.com/google/transit/pull/105#discussion_r222000795, although for some reason @gcamps comment doesn't appear there anymore (?).  It was a suggestion to put a limited time window on experimental fields, and after than window expires they would be adopted (or potentially discarded).  For example, `OccupancyStatus` has been experimental [since 2015](https://groups.google.com/forum/#!msg/gtfs-realtime/_HtNTGp5LxM/23vN1VIHKfkJ).

Second comment at https://github.com/google/transit/pull/105#pullrequestreview-160776233:

>Should there be a definition of consensus? The amendment process is way more formal but it seems to me that if an experimental fields is approved, it becomes hard to change after the fact due to protobuf backward compatibility issues. My proposition would be same process as the spec amendment process but with a "strong majority vote (80%)" instead of unanimous.

@gcamp any further thoughts or proposals on the above?

**Comment by gcamp** [comment]
@barbeau I removed it because I wanted to clean up the timeline of your PR but that didn't work and couldn't revert. That's why it's not showing up. 

I would make the time window for experimental field of 1 year. Hopefully most field should take less time than this. Nothing else significant to add. Strong majority vote still make sense to me.

**Comment by gcamp** [comment]
@barbeau I would agree in principle but in practice we've had big problems integrating producers into the discussion on Github. I've had that with 2 producers who wanted changes on the spec but didn't follow on Github nor put energy on it. (
https://github.com/google/transit/pull/87 and 
https://github.com/google/transit/pull/111)

Maybe that's a symptom of a bigger problem with the producers we are working with, so I would agree.

**Comment by barbeau** [comment]
@gcamp Ok, thanks.  One other question here - after the expiration windows passes, and assuming an experimental field doesn't get adopted, how exactly does it get deprecated?

Looks like our options are, in increasing order of severity:
1. `EXPERIMENTAL_FIELD_THAT_NOBODY_WANTED = 4 [deprecated=true];`
    * [Protobuf docs](https://developers.google.com/protocol-buffers/docs/proto#options) say “code may be annotated as @Deprecated by pb compilers” - I interpret this as just being a hint to the tools, but in theory the field could still be used.
1. `RESERVED 4;`
    * [Protobuf docs](https://developers.google.com/protocol-buffers/docs/proto#reserved) say “The protocol buffer compiler will complain if any future users try to use these field identifiers.” Is this a compiler build failure?  How exactly does it "complain"?

Any thoughts based on experience with protobufs are welcome!

**Comment by gcamp** [comment]
I would prefer option 1., because as much as some field might not be necessary, it would be nice to have the option to revert on using `EXPERIMENTAL_FIELD_THAT_NOBODY_WANTED` to actual use it.

**Comment by abyrd** [comment]
I agree with @barbeau that we should require at least one producer and one consumer for any new feature before it is formally adopted. This is what keeps the spec compact and focused on the concrete domain of passenger information, and prevents ideas from rushing ahead of implementations.

I can imagine as @gcamp said it's difficult to get producers to engage in discussion on Github. But I don't see how this affects the requirement. Requiring real-world implementation before formally modifying the spec seems to be a separate requirement than including the producer in the design discussion.

I have no preference on the experimental field deprecation mechanism. It seems to me that both methods would allow the field to be re-activated in a future version of the spec in the event that the abandoned field is revived a few years later.

**Comment by jakluk** [comment]
Sadly I can't provide any specific suggestions on how to deal with the process, but from my viewpoint the main problem with producers is that they're mostly busy with keeping their live feeds running. I can believe that in most agencies the feeds are maintained by one or two people, whose main responsibilities are completely different. In my case it's solely my free time activity - I was interested in eventually implementing #87, but it was only last week when we were finally ready to publish our data to the world since communication with the internal bus tracking system is not the easiest one.

It means that it's quite demanding to implement experimental changes into the feeds, keep track of changes during the approval process and distribute the extension proto files to the consumers. Another problem is to find consumers who are actually willing to work with the protobuf format, for many developers interested only in data from one agency it's much easier to process plain JSON which they're used to. Extensions or experimental fields just bring complications not only for them.

So basically I guess that our only consumers of protobuf format would be Google Maps and eventually a few more services with worldwide coverage, who can also take quite some time to implement the changes.

**Comment by abyrd** [comment]
Thanks for your comments @jakluk. I do see what you and @gcamp mean about the difficulty of getting producers to contribute to a discussion on the specification or modify their feeds. 

However, I don't really understand how lack of time or resources on the producer side constitutes an argument for or against the requirement to have at least one working producer and consumer before changes to the spec are finalized. It seems only tangentially related to me.

If someone wants the spec to forge ahead quickly with ways to represent all sorts of situations, ahead of people actually implementing them, then they might find that the one-producer-one-consumer requirement slows down progress. But this is precisely the goal of the requirement - to prevent the spec from evolving at a faster rate than existing practice. It is common for people and organizations to generate ideas much faster than they allocate resources to actually understand and test them. The world already has plenty of exhaustive, voluminous data models that no one uses. GTFS benefits from avoiding that path.

**Comment by gcamp** [comment]
@abyrd My experience so far with producers has been that they would be willing to implement something only when it's accepted in the spec, even if it's something that they are asking for. Basically their time is so sparse (like @jakluk said), that they don't want to spend time implementing something that might be rejected later on. 

On top of that, the complexity of having two .proto files simultaneously is quite high, especially for consumer but also for producers. You'll need two proto files to create one compliant GTFS-rt and one experimental GTFS-rt.

**Comment by barbeau** [comment]
@paulswartz Just to clarify, the current [GTFS-realtime change process](https://github.com/google/transit/blob/master/gtfs-realtime/CHANGES.md) encourages producers/consumers to add new fields to the main .proto instead of implementing custom extensions, as long as the field seems generally useful, because of implementation issues.  Just wanted to make sure that was clear for anyone following the discussion.

**Comment by barbeau** [comment]
Proposal for defining GTFS-realtime consensus process based on this discussion is at https://github.com/google/transit/pull/126.

We can tackle the limited window on experimental fields separately.

**Comment by barbeau** [comment]
After talking with someone who has more experience managing .protos long-term, I suggest that we deprecate expired experimental fields using `[deprecated=true]` - so the field would look like:

`EXPERIMENTAL_FIELD_THAT_NOBODY_WANTED = 1 [deprecated=true];`

Here's why:
1. It would allow any existing consumers or producer that have implemented support to leave code as-is.  If we used `RESERVED` instead, consumers or producers would need to remove implemented code as the field would no longer be available the next time the stub code was generated by the protocol buffer compiler.  With `[deprecated=true]`, the stub code may flag the field as deprecated in code annotations, but it would still be usable.
1. It allows easier discovery of past experimental fields when someone wants to implement something similar (again) in the future.

The above would allow us to easily "un-deprecate" a field as well.  For example, let's say that the 2 years expired for an experimental field, and several consumers implemented support but no one is publishing the data yet.  The field would get marked as `[deprecated=true]` in the .proto, but early adopting consumers could leave code that supports the feature in their product.  Then, if a publisher started publishing the data, we could re-adopt the field easily, again without the early adopting consumers needing to change anything.  And then, having the requisite producer/consumer, we could vote to officially adopt the field.

The primary downside to using `[deprecated=true]` instead of `RESERVED` is that the .proto would be more cluttered over time if old expired experimental fields pile up.  But I think the above benefits outweigh the negatives.

Thoughts?

**Comment by barbeau** [comment]
Proposal for producer/consumer requirement, limited time for experimental fields opened at https://github.com/google/transit/pull/140.

**Comment by barbeau** [comment]
https://github.com/google/transit/pull/140 is now approved and merged, and https://github.com/google/transit/pull/126 was previously approved and merged, so this should cover the topics discussed in this issue.  Thanks all!
